durango-cli/src/main.rs

```diff
@@ -0,0 +1,275 @@
+            policy_json.as_str()
+    ) {
+        Ok(policy) => policy,
+        Err(err) => {
+            error!("{}", err);
+            process::exit(1);
+        }
+    };
+    info!("Loaded policy {}", policy_hash);
+
+    // need to convert to str for Durango
+    // TODO allow Durango to accept Paths?
+    let client_cert_path = match opt.identity.to_str() {
+        Some(client_cert_path) => client_cert_path,
+        None => {
+            error!("Invalid client_cert_path (not utf8?)");
+            process::exit(1);
+        }
+    };
+    let client_key_path = match opt.key.to_str() {
+        Some(client_key_path) => client_key_path,
+        None => {
+            error!("Invalid client_key_path (not utf8?)");
+            process::exit(1);
+        }
+    };
+
+    // create Durango instance
+    let mut durango = match Durango::new(
+        client_cert_path,
+        client_key_path,
+        &policy_json,
+    ) {
+        Ok(durango) => durango,
+        Err(err) => {
+            error!("{}", err);
+            process::exit(1);
+        }
+    };
+    info!("Connected to {}", policy.sinaloa_url());
+
+    let mut did_something = false;
+
+    // send program?
+    if let Some(ref program_path) = opt.program {
+        did_something = true;
+
+        let program = if program_path == &path::PathBuf::from("-") {
+            let mut program = Vec::new();
+            match io::stdin().read_to_end(&mut program) {
+                Ok(_) => program,
+                Err(err) => {
+                    error!("{}", err);
+                    process::exit(1);
+                }
+            }
+        } else {
+            match fs::read(program_path) {
+                Ok(program) => program,
+                Err(err) => {
+                    error!("{}", err);
+                    process::exit(1);
+                }
+            }
+        };
+
+        match durango.send_program(&program) {
+            Ok(()) => {}
+            Err(err) => {
+                error!("{}", err);
+                process::exit(1);
+            }
+        }
+
+        info!("Submitted program {:?}", program_path);
+    }
+
+    // send data(s)?
+    // TODO can we specify these in any order?
+    for data_path in opt.data.iter() {
+        did_something = true;
+
+        let data = if data_path == &path::PathBuf::from("-") {
+            let mut data = Vec::new();
+            match io::stdin().read_to_end(&mut data) {
+                Ok(_) => data,
+                Err(err) => {
+                    error!("{}", err);
+                    process::exit(1);
+                }
+            }
+        } else {
+            match fs::read(data_path) {
+                Ok(data) => data,
+                Err(err) => {
+                    error!("{}", err);
+                    process::exit(1);
+                }
+            }
+        };
+
+        match durango.send_data(&data) {
```

Comment:
Maybe rephrase as `if let Err(err) = durango.send_data(&data) {`

Reply:
Did we reach a consensus on this? I do find match statements easier to read for error handling since the error path is the exceptional case.

Reply:
I think we should assess on a case-by-case basis whether `if let` or a `match` statement is more clear.  If you think `match` is clearer here, then we should leave it.